pkg/app/piped/executor/lambda/lambda.go

```diff
@@ -150,17 +152,159 @@ func sync(ctx context.Context, in *executor.Input, cloudProviderName string, clo
+		return false
+	}
+
+	rolloutVersionKeyName := fmt.Sprintf("%s-rollout", fm.Spec.Name)
+	version, ok := in.MetadataStore.Get(rolloutVersionKeyName)
+	if !ok {
+		in.LogPersister.Errorf("Unable to prepare version to promote for Lambda function %s: Not found", fm.Spec.Name)
+		return false
+	}
+
+	options := in.StageConfig.LambdaPromoteStageOptions
+	if options == nil {
+		in.LogPersister.Errorf("Malformed configuration for stage %s", in.Stage.Name)
+		return false
+	}
+
+	trafficCfg, err := client.GetTrafficConfig(ctx, fm)
+	// Create Alias on not yet existed.
+	if errors.Is(err, provider.ErrNotFound) {
+		if options.Percent != 100 {
+			in.LogPersister.Errorf("Not previous version available to handle traffic, new version has to get 100 percent of traffic")
+			return false
+		}
+		if err := client.CreateTrafficConfig(ctx, fm, version); err != nil {
+			in.LogPersister.Errorf("Failed to create traffic routing for Lambda function %s (version: %s): %v", fm.Spec.Name, version, err)
+			return false
+		}
+		in.LogPersister.Infof("Successfully route all traffic to the lambda function %s (version %s)", fm.Spec.Name, version)
+		return true
+	}
+	if err != nil {
+		in.LogPersister.Errorf("Failed to prepare traffic routing for Lambda function %s: %v", fm.Spec.Name, err)
+		return false
+	}
+
+	// Update traffic to the new lambda version.
+	if !configureTrafficRouting(trafficCfg, version, options.Percent) {
+		in.LogPersister.Errorf("Failed to prepare traffic routing for Lambda function %s", fm.Spec.Name)
+		return false
+	}
+
+	if err = client.UpdateTrafficConfig(ctx, fm, trafficCfg); err != nil {
+		in.LogPersister.Errorf("Failed to update traffic routing for Lambda function %s (version: %s): %v", fm.Spec.Name, version, err)
+		return false
+	}
+
+	in.LogPersister.Infof("Successfully promote new version (v%s) of Lambda function %s, it will handle %v percent of traffic", version, fm.Spec.Name, options.Percent)
+	return true
+}
+
+func configureTrafficRouting(trafficCfg provider.RoutingTrafficConfig, version string, percent int) bool {
+	// Update traffic to the new lambda version.
+	primary, ok := trafficCfg["primary"]
+	if !ok {
+		return false
+	}
+	// Set built version by rollout stage as new primary.
+	trafficCfg["primary"] = provider.VersionTraffic{
+		Version: version,
+		Percent: float64(percent),
+	}
+	// Make the current primary version as new secondary version in case it's not the latest built version by rollout stage.
+	if primary.Version != version {
+		trafficCfg["secondary"] = provider.VersionTraffic{
+			Version: primary.Version,
+			Percent: float64(100 - percent),
+		}
+	} else {
+		// Update traffic to the secondary and keep it as new secondary.
+		if secondary, ok := trafficCfg["secondary"]; ok {
+			trafficCfg["secondary"] = provider.VersionTraffic{
+				Version: secondary.Version,
+				Percent: float64(100 - percent),
+			}
+		}
+	}
```

Comment:
How about using `else if` instead of using `if` inside `else` statement?
```suggestion
	} else if secondary, ok := trafficCfg["secondary"]; ok {
		// Update traffic to the secondary and keep it as new secondary.
		trafficCfg["secondary"] = provider.VersionTraffic{
			Version: secondary.Version,
			Percent: float64(100 - percent),
		}
	}
```

Reply:
I get your point 👍 To be honest, I quite don't like that `else if` statement 😓 
The reason is that `else if` is short for
```diff
- } else if (primary.Version == version) && (secondary, ok := trafficCfg["secondary"]); ok {
+ } else if secondary, ok := trafficCfg["secondary"]; ok {
```
The if in else if condition should at least related to the first if to avoid misunderstanding. For example, in case `A v B v C = 1` then `if A else if B else (if C)` is fine.

Reply:
Okay I got your point exactly 👍  We have no explicit coding style guide, but have common ground here: avoid being deeply nested as much as possible.

So how about this?

```go
	// Update traffic to the secondary and keep it as new secondary.
	if secondary, ok := trafficCfg["secondary"]; ok {
		trafficCfg["secondary"] = provider.VersionTraffic{
			Version: secondary.Version,
			Percent: float64(100 - percent),
		}
	}
	// Make the current primary version as new secondary version in case it's not the latest built version by rollout stage.
	if primary.Version != version {
		trafficCfg["secondary"] = provider.VersionTraffic{
			Version: primary.Version,
			Percent: float64(100 - percent),
		}
	}
```

Reply:
In that case, the above (Update traffic to the secondary and keep it as new secondary.) is redundant when the built version is not equal to the current primary version 😂 Edit the `secondary` first, then reset it to another value right after is quite weird to read, I guess 🤔

Reply:
We've actually avoided being deeply nested by doing like that, but I'm unfamiliar with the standard for this case. Besides, it's a matter of taste, I respect your decision 👍

Reply:
Thank you 🙏